src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/DeclareCommands.java

```diff
@@ -0,0 +1,473 @@
+		int i = 0;
+		topLevelNodes = new CommandNode[topLevelNodeIndices.length];
+		for (int index : topLevelNodeIndices) {
+			topLevelNodes[i++] = allNodes[index];
+		}
+	}
+
+	private static CommandNode readCommandNode(ByteBuf from) {
+		byte flags = from.readByte();
+		boolean isPathEnd;
+		int redirect;
+		int[] children = ArraySerializer.readVarIntVarIntArray(from).clone();
+		if ((flags & FLAG_HAS_REDIRECT) != 0) {
+			redirect = VarNumberSerializer.readVarInt(from);
+		} else {
+			redirect = -1;
+		}
+		String name;
+		String argType;
+		String suggestion;
+
+		isPathEnd = ((flags & FLAG_IS_COMMAND_PATH_END) != 0);
+
+		if ((flags & FLAG_IS_LITERAL) != 0) {
+			name = StringSerializer.readVarIntUTF8String(from);
+			argType = null; // no argType signals this is a literal
+			suggestion = null;
+		} else if ((flags & FLAG_IS_ARGUMENT) != 0) {
+			name = StringSerializer.readVarIntUTF8String(from);
+			// This is slightly lossy; some arg types has extra information that we throw away
+			// To properly encode that, we'd need a specialized class, not a string.
+			argType = readArgType(from);
+
+			if ((flags & FLAG_HAS_SUGGESTION) != 0) {
+				suggestion = StringSerializer.readVarIntUTF8String(from);
+			} else {
+				suggestion = null;
+			}
+		} else {
+			// This is only allowed for the root node
+			name = null;
+			argType = null;
+			suggestion = null;
+		}
+		return new CommandNode(name, argType, suggestion, children, redirect, isPathEnd);
+	}
+
+	private static void registerArgTypeReaders() {
+		argTypeReaders.put("brigadier:string", from -> {
+			// Determine kind of string, any of STRING_IS_*...
+			int stringType = VarNumberSerializer.readVarInt(from);
+		});
+		argTypeReaders.put("brigadier:integer", from -> {
+			byte flag = from.readByte();
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				int min = from.readInt();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				int max = from.readInt();
+			}
+		});
+		argTypeReaders.put("brigadier:float", from -> {
+			byte flag = from.readByte();
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				float min = from.readFloat();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				float max = from.readFloat();
+			}
+		});
+		argTypeReaders.put("brigadier:double", from -> {
+			byte flag = from.readByte();
+			if ((flag & FLAG_HAS_MIN_VALUE) != 0) {
+				double min = from.readDouble();
+			}
+			if ((flag & FLAG_HAS_MAX_VALUE) != 0) {
+				double max = from.readDouble();
+			}
+		});
+		argTypeReaders.put("minecraft:entity", from -> {
+			// The flag determines the amount (single or double) and type (players or entities)
+			// See FLAG_ENTITY_AMOUNT_IS_SINGLE and FLAG_ENTITY_TYPE_IS_PLAYER.
+			byte flag = from.readByte();
+		});
+		argTypeReaders.put("minecraft:score_holder", from -> {
+			// The "multiple" boolean is true if multiple, false if single.
+			boolean multiple = from.readBoolean();
+		});
+	}
+
+	private static String readArgType(ByteBuf from) {
```

Comment:
Collapse into single line and move the comment

Reply:
But why?